instance/tests/api/test_openedx_instance.py

```diff
@@ -98,6 +98,48 @@ def assert_active_appserver(self, instance_id, appserver_id):
             )
         return response
 
+    def test_instance_list_efficiency(self):
+        """
+        The number of database queries required to fetch /api/v1/instance/
```

Comment:
@e-kolpakov The main point of this test is *not* to test how many queries are performed when the API is used. The main point of this test is within the `for` loop, which tests that the number of queries stays the same even as the number of instances increases. That is something that we do care about, and it is proportional to the loading speed, as Brandon determined during testing. If the number of queries increases with the number of instances, this API will get slower unnecessarily as the number of instances increases*.

Also, if you look at the queries that are being performed, they are very specific to the overall model we are using, rather than the details that change more often - it's something like 1 query for `InstanceReference`, 1 query for `OpenEdXInstance`, 1 query for `OpenEdXAppServer`, and 1 query for active `OpenEdXAppServer`s (the other queries are standard django things for the current user, session, and/or `content_types`). That basic structure doesn't change frequently.

If we do make changes to the core model in the future, it's totally fine to change `queries_per_api_call = 9` here to some new number - the exact value doesn't matter so much (although I do think it's good to know when it does change). What's most important is that we don't change any other parts of this test.

If we do find that we have to change `queries_per_api_call` a lot, it probably means we have some weird implementation of something that we should fix - alternatively we can modify this test to detect `queries_per_api_call` automatically on the first iteration and then ensure it doesn't change on the 2-3 iterations of the loop - without caring what the exact number is.


\* (unless of course we had some particularly bad data model and queries where the few "optimized" queries have like an O(n³) complexity and the many "unoptimized" queries that get made per instance are only O(√n))